File: sdks/dotnet/src/PrismDb/Update.cs
// Document update operators and their wire encoding.
//
// Mirrors prism_protocol::DocUpdate: an ordered list of $set / $unset / $inc
// mutations. Build with the U helpers. Operand values reuse the tagged Value
// encoding. Carried as the `update` blob of an update command.

using System.Collections.Generic;

namespace PrismDb
{
    /// <summary>One field mutation. Construct via the <see cref="U"/> helpers.</summary>
    public sealed class DocUpdate
    {
        internal string Op = "";
        internal string Field = "";
        internal object? Value;
        internal long Delta;
    }

    /// <summary>Update builders mirroring the engine's update operators.</summary>
    public static class U
    {
        /// <summary>$set - set <paramref name="field"/> to <paramref name="value"/>.</summary>
        public static DocUpdate Set(string field, object? value) =>
            new DocUpdate { Op = "set", Field = field, Value = value };

        /// <summary>$unset - remove <paramref name="field"/>.</summary>
        public static DocUpdate Unset(string field) =>
            new DocUpdate { Op = "unset", Field = field };

        /// <summary>$inc - add <paramref name="delta"/> to the integer <paramref name="field"/>.</summary>
        public static DocUpdate Inc(string field, long delta) =>
            new DocUpdate { Op = "inc", Field = field, Delta = delta };
    }

    internal static class UpdateCodec
    {
        public static byte[] Encode(IReadOnlyList<DocUpdate> ops)
        {
            var w = new Writer();
            w.U32(ops.Count);
            foreach (var op in ops)
            {
                switch (op.Op)
                {
                    case "set":
                        w.U8(1);
                        w.StrU16(op.Field);
                        ValueCodec.EncodeTagged(w, op.Value);
                        break;
                    case "unset":
                        w.U8(2);
                        w.StrU16(op.Field);
                        break;
                    case "inc":
                        w.U8(3);
                        w.StrU16(op.Field);
                        w.I64(op.Delta);
                        break;
                }
            }
            return w.Out();
        }
    }
}

File: sdks/python/prismdb/update.py
"""Document update operators and their wire encoding.

Mirrors ``prism_protocol::DocUpdate``: an ordered list of $set / $unset / $inc
mutations. Build with the :class:`U` helpers. Operand values reuse the tagged
``Value`` encoding. Carried as the ``update`` blob of an update command."""

from __future__ import annotations

from typing import List, Union

from ._codec import Writer
from .value import Value, encode_tagged


class DocUpdate:
    """One field mutation. Construct via the :class:`U` helpers."""

    __slots__ = ("op", "field", "value", "delta")

    def __init__(self, op: str, field: str, value: Value = None, delta: int = 0) -> None:
        self.op = op
        self.field = field
        self.value = value
        self.delta = delta


class U:
    """Update builders mirroring the engine's update operators."""

    @staticmethod
    def set(field: str, value: Value) -> DocUpdate:
        """``$set`` - set ``field`` to ``value``."""
        return DocUpdate("set", field, value=value)

    @staticmethod
    def unset(field: str) -> DocUpdate:
        """``$unset`` - remove ``field``."""
        return DocUpdate("unset", field)

    @staticmethod
    def inc(field: str, delta: Union[int, float]) -> DocUpdate:
        """``$inc`` - add ``delta`` to the integer ``field``."""
        return DocUpdate("inc", field, delta=int(delta))


def encode_doc_update(ops: List[DocUpdate]) -> bytes:
    """Encode a list of update operations to the ``update`` blob of a command."""
    w = Writer()
    w.u32(len(ops))
    for op in ops:
        if op.op == "set":
            w.u8(1)
            w.str_u16(op.field)
            encode_tagged(w, op.value)
        elif op.op == "unset":
            w.u8(2)
            w.str_u16(op.field)
        elif op.op == "inc":
            w.u8(3)
            w.str_u16(op.field)
            w.i64(int(op.delta))
    return w.out()
